service/tasks/driver.py

```diff
@@ -1132,8 +1132,33 @@ def add_floating_ip(driverCls, provider, identity, core_identity_uuid,
                 floating_ip_addr, instance_alias)
             floating_ip = selected_floating_ip
         else:
-            floating_ip = network_driver.associate_floating_ip(instance_alias)
-            floating_ip_addr = floating_ip["floating_ip_address"]
+            if core_identity.provider.cloud_config['network']['topology'] \
+                    == "External Router Topology":
+                # Determine correct external network based on external gateway
+                # info of the identity's public router
+                public_router_name = identity.get_credential('router_name')
+                admin_identity = core_identity.provider.admin
+                admin_neutron = instance_service._to_network_driver(admin_identity).neutron
+                routers = admin_neutron.list_routers(retrieve_all=True)['routers']
+                public_router = None
+                for router in routers:
+                    if router['name'] == public_router_name:
+                        public_router = router
+                if not public_router:
+                    raise Exception("Could not find a router matching"
+                                    " public_router name {} for user {}"
```

Comment:
I made a small change to remove the specific numeric reference inside each `{}` which is only primarily useful when you want them to appear in a different order than listed, or more than once.